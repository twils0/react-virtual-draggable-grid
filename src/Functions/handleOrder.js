import findMaxHeight from './findMaxHeight';
import findMaxWidth from './findMaxWidth';
import handleOrderObject from './handleOrderObject';
import getPositionLeft from './getPositionLeft';
import getPositionTop from './getPositionTop';

const handleOrder = ({
  items,
  order,
  keys,
  initialSizeBool,
  fixedRows,
  fixedColumns,
  fixedWidthAll,
  fixedHeightAll,
  gutterX,
  gutterY,
}) => {
  const newOrder = [];
  const newKeys = {};
  let resizeNeededCount = 0;
  let maxBottom = null;
  const maxRightArray = [];

  if (fixedColumns && (!fixedWidthAll || !fixedHeightAll)) {
    maxRightArray[0] = 0;
  }

  if (fixedRows && (!fixedWidthAll || !fixedHeightAll)) {
    maxBottom = 0;
  }

  for (let iY = 0; iY < items.length; iY += 1) {
    const itemsRow = items[iY];

    if (fixedRows && iY > 0 && (!fixedWidthAll || !fixedHeightAll)) {
      maxBottom += fixedHeightAll || findMaxHeight({ order: newOrder, indexY: iY - 1 });
    }

    if (itemsRow && Array.isArray(itemsRow)) {
      const row = [];

      newOrder.push(row);

      for (let iX = 0; iX < itemsRow.length; iX += 1) {
        const item = itemsRow[iX];

        const orderObject = handleOrderObject({
          order,
          keys,
          item,
          row,
          indexX: iX,
          indexY: iY,
          fixedWidthAll,
          fixedHeightAll,
        });

        row.push(orderObject);
        newKeys[item.key] = orderObject;

        if (
          fixedColumns
          && typeof maxRightArray[iX] !== 'number'
          && (!fixedWidthAll || !fixedHeightAll)
        ) {
          // previous order should be available, unless grid has just mounted;
          // need previous order to see the entire column
          maxRightArray[iX] = maxRightArray[iX - 1]
            + (fixedWidthAll
              || findMaxWidth({
                order: order || newOrder,
                indexX: iX - 1,
              }));
        }

        if (!initialSizeBool) {
          if (!fixedWidthAll || !fixedHeightAll) {
            if (!fixedWidthAll) {
              if (typeof maxRightArray[iX] === 'number') {
                orderObject.left = maxRightArray[iX] > 0 ? maxRightArray[iX] + gutterX : maxRightArray[iX];
              } else {
                orderObject.left = getPositionLeft({
                  order: newOrder,
                  orderX: iX,
                  orderY: iY,
                  gutterX,
                });
              }
            } else {
              orderObject.left = iX * (fixedWidthAll + gutterX);
            }
          }

          if (!fixedWidthAll || !fixedHeightAll) {
            if (!fixedHeightAll) {
              if (typeof maxBottom === 'number') {
                orderObject.top = maxBottom > 0 ? maxBottom + gutterY : maxBottom;
              } else {
                orderObject.top = getPositionTop({
                  order: newOrder,
                  keys: newKeys,
                  orderX: iX,
                  orderY: iY,
                  gutterX,
                  gutterY,
                });
              }
            } else {
              orderObject.top = iY * (fixedHeightAll + gutterY);
            }
          }
        } else {
          const { fixedWidth, fixedHeight } = item;

          if (!fixedWidth || !fixedHeight) {
            resizeNeededCount += 1;
          }
        }
      }
    } else {
      const orderObject = handleOrderObject({
        order,
        keys,
        item: itemsRow,
        indexX: 0,
        indexY: iY,
        fixedWidthAll,
        fixedHeightAll,
      });

      newOrder.push([orderObject]);
      newKeys[itemsRow.key] = orderObject;

      if (!initialSizeBool) {
        if (!fixedWidthAll || !fixedHeightAll) {
          orderObject.left = 0;
        }

        if (!fixedWidthAll || !fixedHeightAll) {
          if (!fixedHeightAll) {
            if (typeof maxBottom === 'number') {
              orderObject.top = maxBottom > 0 ? maxBottom + gutterY : maxBottom;
            } else {
              orderObject.top = getPositionTop({
                order: newOrder,
                keys: newKeys,
                orderX: 0,
                orderY: iY,
                gutterX,
                gutterY,
              });
            }
          } else {
            orderObject.top = iY * (fixedHeightAll + gutterY);
          }
        }
      }
    }
  }

  return { order: newOrder, keys: newKeys, resizeNeededCount };
};

export default handleOrder;
