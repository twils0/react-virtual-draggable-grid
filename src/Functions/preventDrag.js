// prevent unwanted 'ghost' drag effect in browser
const preventDrag = (event) => {
  event.preventDefault();
  return false;
};

export default preventDrag;
