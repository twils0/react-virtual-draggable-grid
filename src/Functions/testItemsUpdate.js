const testItemsUpdate = ({ items, order, keys }) => {
  if (!order || !keys) {
    return true;
  }

  const keysLen = Object.keys(keys).length;
  let count = 0;

  const someBoolItems = items.some((itemsRow, indexY) => {
    if (itemsRow && Array.isArray(itemsRow)) {
      const someBoolRow = itemsRow.some((item, indexX) => {
        if (item && item.key && typeof item.key === 'string') {
          const orderObject = keys[item.key];

          if (!orderObject) {
            return true;
          }

          const { itemX, itemY } = orderObject;

          if (itemX !== indexX || itemY !== indexY) {
            return true;
          }

          count += 1;
        }

        return false;
      });

      return someBoolRow;
    }

    if (itemsRow && itemsRow.key) {
      const orderObject = keys[itemsRow.key];

      if (!orderObject) {
        return true;
      }

      const { itemX, itemY } = orderObject;

      if (itemX !== 0 || itemY !== indexY) {
        return true;
      }

      count += 1;
    }

    return false;
  });

  if (!someBoolItems && count !== keysLen) {
    return true;
  }

  return someBoolItems;
};

export default testItemsUpdate;
