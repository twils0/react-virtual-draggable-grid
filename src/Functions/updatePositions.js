import findMaxHeight from './findMaxHeight';
import findMaxWidth from './findMaxWidth';
import getPositionLeft from './getPositionLeft';
import getPositionTop from './getPositionTop';

const updatePositions = ({
  order,
  keys,
  fixedRows,
  fixedColumns,
  fixedWidthAll,
  fixedHeightAll,
  orderX,
  orderY,
  gutterX,
  gutterY,
}) => {
  const newOrder = [...order];
  const newKeys = { ...keys };
  const limitedOrderY = orderY > 0 ? orderY - 1 : 0;
  const startingY = fixedRows ? 0 : limitedOrderY;
  const firstStartingX = fixedColumns ? 0 : orderX;
  let maxBottom = null;
  const maxRightArray = [];

  if (fixedColumns && (!fixedWidthAll || !fixedHeightAll)) {
    maxRightArray[0] = 0;
  }

  if (fixedRows && (!fixedWidthAll || !fixedHeightAll)) {
    maxBottom = 0;
  }

  for (let iY = startingY; iY < newOrder.length; iY += 1) {
    const orderRow = newOrder[iY];
    const newOrderRow = [...orderRow];
    const startingX = iY === startingY ? firstStartingX : 0;

    newOrder[iY] = newOrderRow;

    if (fixedRows && iY > 0 && (!fixedWidthAll || !fixedHeightAll)) {
      maxBottom += fixedHeightAll || findMaxHeight({ order: newOrder, indexY: iY - 1 });
    }

    for (let iX = startingX; iX < newOrderRow.length; iX += 1) {
      const orderObject = newOrderRow[iX];
      const newOrderObject = { ...orderObject };

      if (
        fixedColumns
        && typeof maxRightArray[iX] !== 'number'
        && (!fixedWidthAll || !fixedHeightAll)
      ) {
        maxRightArray[iX] = maxRightArray[iX - 1]
          + (fixedWidthAll
            || findMaxWidth({
              order: newOrder,
              indexX: iX - 1,
            }));
      }

      newOrderRow[iX] = newOrderObject;
      newKeys[newOrderObject.key] = newOrderObject;

      newOrderObject.orderX = iX;
      newOrderObject.orderY = iY;

      if (!fixedWidthAll || !fixedHeightAll) {
        if (!fixedWidthAll) {
          if (typeof maxRightArray[iX] === 'number') {
            newOrderObject.left = maxRightArray[iX] > 0 ? maxRightArray[iX] + gutterX : maxRightArray[iX];
          } else {
            newOrderObject.left = getPositionLeft({
              order: newOrder,
              orderX: iX,
              orderY: iY,
              gutterX,
            });
          }
        } else {
          newOrderObject.left = iX * (fixedWidthAll + gutterX);
        }
      }

      if (!fixedWidthAll || !fixedHeightAll) {
        if (!fixedHeightAll) {
          if (typeof maxBottom === 'number') {
            newOrderObject.top = maxBottom > 0 ? maxBottom + gutterY : maxBottom;
          } else {
            newOrderObject.top = getPositionTop({
              order: newOrder,
              keys: newKeys,
              orderX: iX,
              orderY: iY,
              gutterX,
              gutterY,
            });
          }
        } else {
          newOrderObject.top = iY * (fixedHeightAll + gutterY);
        }
      }
    }
  }

  return { order: newOrder, keys: newKeys };
};

export default updatePositions;
