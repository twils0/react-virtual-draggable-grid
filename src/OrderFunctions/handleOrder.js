import handlePosition from './handlePosition';
import getOrderIndex from './getOrderIndex';

const handleOrder = ({ items, order }) => {
  const keys = [];
  const newOrder = [];

  items.forEach((itemY, indexY) => {
    if (itemY && Array.isArray(itemY)) {
      const row = [];
      newOrder.push(row);

      // test if item exists, is an object, and has a key that's
      // not already been used
      itemY.forEach((itemX, indexX) => {
        if (
          itemX
          && typeof itemX === 'object'
          && typeof itemX.ItemComponent === 'function'
          && (itemX.key !== null || itemX.key !== undefined)
          && keys.indexOf(itemX.key) === -1
        ) {
          const {
            key, fixedWidth, fixedHeight, estimatedWidth, estimatedHeight,
          } = itemX;
          let orderWidth = null;
          let orderHeight = null;

          if (order) {
            const { orderIndexX, orderIndexY } = getOrderIndex(order, key);
            const orderRow = order[orderIndexY];

            if (orderRow) {
              const orderObject = orderRow[orderIndexX];

              if (orderObject) {
                orderWidth = orderObject.width;
                orderHeight = orderObject.height;
              }
            }
          }

          // use either fixed size, cached order object size,
          // estimated size, or the default value: 100
          const width = fixedWidth || orderWidth || estimatedWidth || 100;
          const height = fixedHeight || orderHeight || estimatedHeight || 100;
          const orderObject = {
            key,
            itemIndexX: indexX,
            itemIndexY: indexY,
            width,
            height,
          };

          row.push(orderObject);

          // location of the top left corner of the item
          const { left, top } = handlePosition({
            order: newOrder,
            orderIndexX: indexX,
            orderIndexY: indexY,
          });

          keys.push(key);
          orderObject.left = left;
          orderObject.top = top;
        }
      });

      // test if item exists, is an object, and has a key that's
      // not already been used
    } else if (
      itemY
      && typeof itemY === 'object'
      && typeof itemY.ItemComponent === 'function'
      && (itemY.key !== null || itemY.key !== undefined)
      && keys.indexOf(itemY.key) === -1
    ) {
      const {
        key, fixedWidth, fixedHeight, estimatedWidth, estimatedHeight,
      } = itemY;
      let orderWidth = null;
      let orderHeight = null;

      if (order) {
        const { orderIndexX, orderIndexY } = getOrderIndex(order, key);
        const orderRow = order[orderIndexY];

        if (orderRow) {
          const orderObject = orderRow[orderIndexX];

          if (orderObject) {
            orderWidth = orderObject.width;
            orderHeight = orderObject.height;
          }
        }
      }

      // use either fixedWidth, estimatedWidth, or the default value: 100
      const width = fixedWidth || orderWidth || estimatedWidth || 100;
      const height = fixedHeight || orderHeight || estimatedHeight || 100;
      const orderObject = {
        key,
        itemIndexX: 0,
        itemIndexY: indexY,
        width,
        height,
      };

      newOrder.push([orderObject]);

      const { left, top } = handlePosition({
        order: newOrder,
        orderIndexX: 0,
        orderIndexY: indexY,
      });

      keys.push(key);
      orderObject.left = left;
      orderObject.top = top;
    }
  });

  return newOrder;
};

export default handleOrder;
