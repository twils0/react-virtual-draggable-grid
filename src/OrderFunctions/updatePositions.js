import handlePosition from './handlePosition';

const updatePositions = ({
  order, width, height, indexX, indexY,
}) => {
  console.log('up pos 1', indexX, indexY);

  for (let iY = indexY; iY < order.length; iY += 1) {
    for (let iX = indexX; iX < order[iY].length; iX += 1) {
      const orderObject = iX === indexX && iY === indexY
        ? {
          ...order[iY][iX],
          width,
          height,
        }
        : order[iY][iX];

      const { left, top } = handlePosition({
        order,
        width: orderObject.width,
        height: orderObject.height,
        indexX: iX,
        indexY: iY,
      });

      order[iY][iX] = { ...orderObject, left, top }; // eslint-disable-line
    }
  }

  return order;
};

export default updatePositions;
