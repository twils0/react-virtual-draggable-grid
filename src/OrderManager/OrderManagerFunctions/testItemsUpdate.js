// test if items 1D or 2D array has changed
const testItemsUpdate = ({
  items,
  order,
  keys,
  fixedWidthAll,
  fixedHeightAll,
}) => {
  if (!order || !keys) {
    return true;
  }

  const keysLen = Object.keys(keys).length;
  let count = 0;

  const someBoolItems = items.some((itemsRow, indexY) => {
    if (itemsRow && Array.isArray(itemsRow)) {
      const someBoolRow = itemsRow.some((item, indexX) => {
        // ignore any item object missing an ItemComponent,
        // key, fixedWidth, or fixedHeight prop
        if (
          item
          && typeof item === 'object'
          && typeof item.ItemComponent === 'function'
          && typeof item.key === 'string'
          && typeof item.fixedWidth === 'number'
          && typeof item.fixedHeight === 'number'
        ) {
          const orderObject = keys[item.key];

          // if there's not an orderObject for the item, it's new
          if (!orderObject) {
            return true;
          }

          const {
            itemX, itemY, width, height,
          } = orderObject;

          // if the orderObject's item indexes do not
          // match the indexes from the items 1D or 2D array,
          // something has changed
          if (itemX !== indexX || itemY !== indexY) {
            return true;
          }

          if (
            (width !== item.fixedWidth && width !== fixedWidthAll)
            || (height !== item.fixedHeight && height !== fixedHeightAll)
          ) {
            return true;
          }

          // count the number of items provided
          count += 1;
        }

        return false;
      });

      return someBoolRow;
    }

    // ignore items without a key prop
    if (itemsRow && itemsRow.key) {
      const orderObject = keys[itemsRow.key];

      // if there's not an orderObject for the item, it's new
      if (!orderObject) {
        return true;
      }

      const { itemX, itemY } = orderObject;

      // if the orderObject's item indexes do not
      // match the indexes from the items 1D or 2D array,
      // something has changed
      if (itemX !== 0 || itemY !== indexY) {
        return true;
      }

      // count the number of items provided
      count += 1;
    }

    return false;
  });

  // if the number of items provided doesn't match the
  // current number of object keys in the keys object, something
  // has changed
  if (!someBoolItems && count !== keysLen) {
    return true;
  }

  return someBoolItems;
};

export default testItemsUpdate;
