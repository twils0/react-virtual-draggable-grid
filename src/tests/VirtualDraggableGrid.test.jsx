import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';

import VirtualDraggableGrid from '../VirtualDraggableGrid';

import handleOrder from '../Functions/handleOrder';
import updatePositions from '../Functions/updatePositions';
import testItemsUpdate from '../Functions/testItemsUpdate';
import preventDrag from '../Utilities/preventDrag';

jest.mock('../Functions/handleOrder', () => jest.fn());
jest.mock('../Functions/updatePositions', () => jest.fn());
jest.mock('../Functions/testItemsUpdate', () => jest.fn());
jest.mock('../Utilities/preventDrag', () => jest.fn());

const shallowComponent = props => shallow(<VirtualDraggableGrid {...props} />);

const TestComp = (props) => {
  const { styles, name } = props;

  return (
    <div
      style={{
        display: 'flex',
        justifyContent: 'center',
        alignText: 'center',
        userSelect: 'none',
        width: '50px',
        height: '50px',
        backgroundColor: 'lightblue',
        border: '1px solid black',
        ...styles,
      }}
    >
      {`Test ${name}`}
    </div>
  );
};

TestComp.propTypes = {
  styles: PropTypes.object,
  name: PropTypes.string.isRequired,
};

TestComp.defaultProps = {
  styles: {},
};

const defaultProps = {
  items: [
    [
      {
        key: 'test-0',
        ItemComponent: TestComp,
        fixedWidth: 100,
        fixedHeight: 100,
        itemProps: {
          name: 'test-0',
          style: { userSelect: 'none', width: 100, height: 100 },
        },
      },
      {
        key: 'test-1',
        ItemComponent: TestComp,
        fixedWidth: 200,
        fixedHeight: 200,
        itemProps: {
          name: 'test-1',
          style: { userSelect: 'none', width: 200, height: 200 },
        },
      },
    ],
    [
      {
        key: 'test-2',
        ItemComponent: TestComp,
        fixedWidth: 300,
        fixedHeight: 300,
        itemProps: {
          name: 'test-2',
          style: { userSelect: 'none', width: 300, height: 300 },
        },
      },
      {
        key: 'test-3',
        ItemComponent: TestComp,
        fixedWidth: 400,
        fixedHeight: 400,
        itemProps: {
          name: 'test-3',
          style: { userSelect: 'none', width: 400, height: 400 },
        },
      },
    ],
    {
      key: 'test-4',
      ItemComponent: TestComp,
      fixedWidth: 500,
      fixedHeight: 500,
      itemProps: {
        name: 'test-4',
        style: { userSelect: 'none', width: 500, height: 500 },
      },
    },
  ],
  fixedRows: false,
  fixedColumns: false,
  fixedWidthAll: null,
  fixedHeightAll: null,
  onlyDragElements: [],
  onlyDragIds: [],
  noDragElements: [],
  noDragIds: [],
  gutterX: 0,
  gutterY: 0,
  mouseUpdateTime: 100,
  mouseUpdateX: 50,
  mouseUpdateY: 50,
  leeway: 0.1,
  scrollBufferX: 200,
  scrollBufferY: 200,
  scrollUpdateX: 100,
  scrollUpdateY: 200,
  transitionDuration: '0.3s',
  transitionTimingFunction: 'ease',
  transitionDelay: '0.2s',
  shadowMultiple: 16,
  shadowHRatio: 1,
  shadowVRatio: 1,
  shadowBlur: null,
  shadowBlurRatio: 1.2,
  shadowSpread: null,
  shadowSpreadRatio: 0,
  shadowColor: 'rgba(0, 0, 0, 0.2)',
  WrapperStyles: {},
  GridStyles: {},
  GridItemStyles: {},
  getItems: jest.fn(),
  getVisibleItems: jest.fn(),
};

const constructorState = {
  order: [
    [
      {
        key: 'test-0',
        itemX: 0,
        itemY: 0,
        orderX: 0,
        orderY: 0,
        width: 100,
        height: 100,
        left: 0,
        top: 0,
      },
      {
        key: 'test-1',
        itemX: 1,
        itemY: 0,
        orderX: 1,
        orderY: 0,
        width: 200,
        height: 200,
        left: 100,
        top: 0,
      },
    ],
    [
      {
        key: 'test-2',
        itemX: 0,
        itemY: 1,
        orderX: 0,
        orderY: 1,
        width: 300,
        height: 300,
        left: 0,
        top: 200,
      },
      {
        key: 'test-3',
        itemX: 1,
        itemY: 1,
        orderX: 1,
        orderY: 1,
        width: 400,
        height: 400,
        left: 300,
        top: 0,
      },
    ],
    [
      {
        key: 'test-4',
        itemX: 0,
        itemY: 2,
        orderX: 0,
        orderY: 2,
        width: 500,
        height: 500,
        left: 0,
        top: 500,
      },
    ],
  ],
  itemsBool: true,
  fixedColumns: false,
  fixedHeightAll: null,
  fixedRows: false,
  fixedWidthAll: null,
  gutterX: 0,
  gutterY: 0,
  leeway: 0.1,
  scrollBufferX: 200,
  scrollBufferY: 200,
};

constructorState.keys = {
  'test-0': constructorState.order[0][0],
  'test-1': constructorState.order[0][1],
  'test-2': constructorState.order[1][0],
  'test-3': constructorState.order[1][1],
  'test-4': constructorState.order[2][0],
};

const updatedOrder = [
  ...constructorState.order,
  [
    {
      key: 'test-5',
      itemX: 0,
      itemY: 3,
      orderX: 0,
      orderY: 3,
      width: 600,
      height: 600,
      left: 0,
      top: 1000,
    },
  ],
];
const updatedKeys = {
  ...constructorState.keys,
  'test-5': updatedOrder[3][0],
};

describe('VirtualDraggableGrid', () => {
  afterEach(() => {
    defaultProps.getItems.mockReset();
    handleOrder.mockReset();
    updatePositions.mockReset();
    testItemsUpdate.mockReset();
    preventDrag.mockReset();
  });

  it('renders correctly', () => {
    const { items } = defaultProps;
    const { order, keys } = constructorState;

    handleOrder.mockReturnValue({ order, keys });

    const wrapper = shallowComponent(defaultProps);
    const instance = wrapper.instance();
    const instanceProps = instance.props;

    const instanceState = instance.state;

    expect(instanceProps).toEqual(defaultProps);
    expect(instanceState).toEqual(constructorState);
    expect(handleOrder).toHaveBeenCalledTimes(1);
    expect(handleOrder).toBeCalledWith({
      items,
      fixedRows: false,
      fixedColumns: false,
      fixedWidthAll: null,
      fixedHeightAll: null,
      gutterX: 0,
      gutterY: 0,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('getDerivedStateFromProps executes correctly, items not updated', () => {
    const { items } = defaultProps;
    const { order, keys } = constructorState;

    testItemsUpdate.mockReturnValue(false);
    handleOrder.mockReturnValueOnce({ order, keys });

    const wrapper = shallowComponent(defaultProps);

    const instance = wrapper.instance();
    const instanceProps = instance.props;
    const instanceState = instance.state;

    expect(instanceProps).toEqual(defaultProps);
    expect(instanceState).toEqual(constructorState);
    expect(handleOrder).toHaveBeenCalledTimes(1);
    expect(handleOrder).toBeCalledWith({
      items,
      fixedRows: false,
      fixedColumns: false,
      fixedWidthAll: null,
      fixedHeightAll: null,
      gutterX: 0,
      gutterY: 0,
    });
  });

  it('getDerivedStateFromProps executes correctly, items updated', () => {
    const { items } = defaultProps;
    const { order, keys } = constructorState;
    const updatedState = {
      ...constructorState,
      order: updatedOrder,
      keys: updatedKeys,
    };

    testItemsUpdate.mockReturnValueOnce(true);
    testItemsUpdate.mockReturnValue(false);
    handleOrder.mockReturnValueOnce({
      order,
      keys,
    });
    handleOrder.mockReturnValue({
      order: updatedOrder,
      keys: updatedKeys,
    });

    const wrapper = shallowComponent(defaultProps);

    const instance = wrapper.instance();
    const instanceProps = instance.props;
    const instanceState = instance.state;

    expect(instanceProps).toEqual(defaultProps);
    expect(instanceState).toEqual(updatedState);
    expect(handleOrder).toHaveBeenCalledTimes(2);
    expect(handleOrder).toHaveBeenNthCalledWith(1, {
      items,
      fixedRows: false,
      fixedColumns: false,
      fixedWidthAll: null,
      fixedHeightAll: null,
      gutterX: 0,
      gutterY: 0,
    });
    expect(handleOrder).toHaveBeenNthCalledWith(2, {
      items,
      fixedRows: false,
      fixedColumns: false,
      fixedWidthAll: null,
      fixedHeightAll: null,
      gutterX: 0,
      gutterY: 0,
    });
  });

  it('updateOrder executes correctly', () => {
    const { order, keys } = constructorState;
    const updatedState = {
      ...constructorState,
      order: updatedOrder,
      keys: updatedKeys,
    };

    handleOrder.mockReturnValue({ order, keys });

    const wrapper = shallowComponent(defaultProps);
    const instance = wrapper.instance();
    const instanceProps = instance.props;

    instance.updateOrder({ order: updatedOrder, keys: updatedKeys });

    const instanceState = instance.state;

    expect(instanceProps).toEqual(defaultProps);
    expect(instanceState).toEqual(updatedState);
  });

  it('updateItems executes correctly', () => {
    const { items } = defaultProps;
    const { order, keys } = constructorState;
    const movedItems = [...items];
    movedItems[1] = [...items[1]];
    movedItems[2] = [items[2]];
    const movedOrder = [...order];
    movedOrder[1] = [...order[1]];
    movedOrder[2] = [...order[2]];
    const movedKeys = { ...keys };

    const orderObject = movedOrder[1].pop();
    movedKeys[orderObject.key] = { orderX: 1, orderY: 2 };
    movedOrder[2].push(orderObject);

    const item = movedItems[1].pop();
    [, [movedItems[1]]] = movedItems;
    movedItems[2].push(item);

    const updatedState = {
      ...constructorState,
      order: movedOrder,
      keys: movedKeys,
    };

    testItemsUpdate.mockReturnValue(false);
    handleOrder.mockReturnValue({ order: movedOrder, keys: movedKeys });

    const wrapper = shallowComponent(defaultProps);
    const instance = wrapper.instance();
    const instanceProps = instance.props;

    instance.updateItems();

    const instanceState = instance.state;

    expect(instanceProps).toEqual(defaultProps);
    expect(instanceState).toEqual(updatedState);
    expect(defaultProps.getItems).toHaveBeenCalledTimes(1);
    expect(defaultProps.getItems).toBeCalledWith(movedItems);
  });
});
