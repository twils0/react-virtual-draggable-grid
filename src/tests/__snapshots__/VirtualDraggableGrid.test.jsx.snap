// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualDraggableGrid renders correctly 1`] = `
<div
  className="rvdl-wrapper"
  onDragStart={[MockFunction]}
  style={
    Object {
      "MozUserSelect": "none",
      "boxSizing": "border-box",
      "display": "block",
      "height": "100%",
      "margin": 0,
      "padding": 0,
      "position": "relative",
      "userSelect": "none",
      "width": "100%",
    }
  }
>
  <Grid
    GridItemStyles={Object {}}
    GridStyles={Object {}}
    fixedColumns={false}
    fixedHeightAll={null}
    fixedRows={false}
    fixedWidthAll={null}
    gutterX={0}
    gutterY={0}
    items={
      Array [
        Array [
          Object {
            "ItemComponent": [Function],
            "fixedHeight": 100,
            "fixedWidth": 100,
            "itemProps": Object {
              "name": "test-0",
              "style": Object {
                "height": 100,
                "userSelect": "none",
                "width": 100,
              },
            },
            "key": "test-0",
          },
          Object {
            "ItemComponent": [Function],
            "fixedHeight": 200,
            "fixedWidth": 200,
            "itemProps": Object {
              "name": "test-1",
              "style": Object {
                "height": 200,
                "userSelect": "none",
                "width": 200,
              },
            },
            "key": "test-1",
          },
        ],
        Array [
          Object {
            "ItemComponent": [Function],
            "fixedHeight": 300,
            "fixedWidth": 300,
            "itemProps": Object {
              "name": "test-2",
              "style": Object {
                "height": 300,
                "userSelect": "none",
                "width": 300,
              },
            },
            "key": "test-2",
          },
          Object {
            "ItemComponent": [Function],
            "fixedHeight": 400,
            "fixedWidth": 400,
            "itemProps": Object {
              "name": "test-3",
              "style": Object {
                "height": 400,
                "userSelect": "none",
                "width": 400,
              },
            },
            "key": "test-3",
          },
        ],
        Object {
          "ItemComponent": [Function],
          "fixedHeight": 500,
          "fixedWidth": 500,
          "itemProps": Object {
            "name": "test-4",
            "style": Object {
              "height": 500,
              "userSelect": "none",
              "width": 500,
            },
          },
          "key": "test-4",
        },
      ]
    }
    keys={
      Object {
        "test-0": Object {
          "height": 100,
          "itemX": 0,
          "itemY": 0,
          "key": "test-0",
          "left": 0,
          "orderX": 0,
          "orderY": 0,
          "top": 0,
          "width": 100,
        },
        "test-1": Object {
          "height": 200,
          "itemX": 1,
          "itemY": 0,
          "key": "test-1",
          "left": 100,
          "orderX": 1,
          "orderY": 0,
          "top": 0,
          "width": 200,
        },
        "test-2": Object {
          "height": 300,
          "itemX": 0,
          "itemY": 1,
          "key": "test-2",
          "left": 0,
          "orderX": 0,
          "orderY": 1,
          "top": 200,
          "width": 300,
        },
        "test-3": Object {
          "height": 400,
          "itemX": 1,
          "itemY": 1,
          "key": "test-3",
          "left": 300,
          "orderX": 1,
          "orderY": 1,
          "top": 0,
          "width": 400,
        },
        "test-4": Object {
          "height": 500,
          "itemX": 0,
          "itemY": 2,
          "key": "test-4",
          "left": 0,
          "orderX": 0,
          "orderY": 2,
          "top": 500,
          "width": 500,
        },
      }
    }
    leeway={0.1}
    mouseUpdateTime={100}
    mouseUpdateX={50}
    mouseUpdateY={50}
    noDragElements={Array []}
    noDragIds={Array []}
    onDragStart={[MockFunction]}
    onlyDragElements={Array []}
    onlyDragIds={Array []}
    order={
      Array [
        Array [
          Object {
            "height": 100,
            "itemX": 0,
            "itemY": 0,
            "key": "test-0",
            "left": 0,
            "orderX": 0,
            "orderY": 0,
            "top": 0,
            "width": 100,
          },
          Object {
            "height": 200,
            "itemX": 1,
            "itemY": 0,
            "key": "test-1",
            "left": 100,
            "orderX": 1,
            "orderY": 0,
            "top": 0,
            "width": 200,
          },
        ],
        Array [
          Object {
            "height": 300,
            "itemX": 0,
            "itemY": 1,
            "key": "test-2",
            "left": 0,
            "orderX": 0,
            "orderY": 1,
            "top": 200,
            "width": 300,
          },
          Object {
            "height": 400,
            "itemX": 1,
            "itemY": 1,
            "key": "test-3",
            "left": 300,
            "orderX": 1,
            "orderY": 1,
            "top": 0,
            "width": 400,
          },
        ],
        Array [
          Object {
            "height": 500,
            "itemX": 0,
            "itemY": 2,
            "key": "test-4",
            "left": 0,
            "orderX": 0,
            "orderY": 2,
            "top": 500,
            "width": 500,
          },
        ],
      ]
    }
    orderManager={
      Object {
        "setOrder": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Object {
                "keys": Object {
                  "test-0": Object {
                    "height": 100,
                    "itemX": 0,
                    "itemY": 0,
                    "key": "test-0",
                    "left": 0,
                    "orderX": 0,
                    "orderY": 0,
                    "top": 0,
                    "width": 100,
                  },
                  "test-1": Object {
                    "height": 200,
                    "itemX": 1,
                    "itemY": 0,
                    "key": "test-1",
                    "left": 100,
                    "orderX": 1,
                    "orderY": 0,
                    "top": 0,
                    "width": 200,
                  },
                  "test-2": Object {
                    "height": 300,
                    "itemX": 0,
                    "itemY": 1,
                    "key": "test-2",
                    "left": 0,
                    "orderX": 0,
                    "orderY": 1,
                    "top": 200,
                    "width": 300,
                  },
                  "test-3": Object {
                    "height": 400,
                    "itemX": 1,
                    "itemY": 1,
                    "key": "test-3",
                    "left": 300,
                    "orderX": 1,
                    "orderY": 1,
                    "top": 0,
                    "width": 400,
                  },
                  "test-4": Object {
                    "height": 500,
                    "itemX": 0,
                    "itemY": 2,
                    "key": "test-4",
                    "left": 0,
                    "orderX": 0,
                    "orderY": 2,
                    "top": 500,
                    "width": 500,
                  },
                },
                "order": Array [
                  Array [
                    Object {
                      "height": 100,
                      "itemX": 0,
                      "itemY": 0,
                      "key": "test-0",
                      "left": 0,
                      "orderX": 0,
                      "orderY": 0,
                      "top": 0,
                      "width": 100,
                    },
                    Object {
                      "height": 200,
                      "itemX": 1,
                      "itemY": 0,
                      "key": "test-1",
                      "left": 100,
                      "orderX": 1,
                      "orderY": 0,
                      "top": 0,
                      "width": 200,
                    },
                  ],
                  Array [
                    Object {
                      "height": 300,
                      "itemX": 0,
                      "itemY": 1,
                      "key": "test-2",
                      "left": 0,
                      "orderX": 0,
                      "orderY": 1,
                      "top": 200,
                      "width": 300,
                    },
                    Object {
                      "height": 400,
                      "itemX": 1,
                      "itemY": 1,
                      "key": "test-3",
                      "left": 300,
                      "orderX": 1,
                      "orderY": 1,
                      "top": 0,
                      "width": 400,
                    },
                  ],
                  Array [
                    Object {
                      "height": 500,
                      "itemX": 0,
                      "itemY": 2,
                      "key": "test-4",
                      "left": 0,
                      "orderX": 0,
                      "orderY": 2,
                      "top": 500,
                      "width": 500,
                    },
                  ],
                ],
              },
            },
          ],
        },
        "testItemsUpdate": [MockFunction],
        "updateVisibleOrderNoState": [MockFunction],
      }
    }
    scrollBufferX={200}
    scrollBufferY={200}
    scrollUpdateX={100}
    scrollUpdateY={200}
    shadowBlur={null}
    shadowBlurRatio={1.2}
    shadowColor="rgba(0, 0, 0, 0.2)"
    shadowHRatio={1}
    shadowMultiple={16}
    shadowSpread={null}
    shadowSpreadRatio={0}
    shadowVRatio={1}
    transitionDelay="0.2s"
    transitionDuration="0.3s"
    transitionTimingFunction="ease"
    visibleOrder={Array []}
  />
</div>
`;
